docs/core/howto/logger.rst

```diff
@@ -202,7 +202,7 @@ Format strings
 ~~~~~~~~~~~~~~
 
 
-Format strings provide observers with a standard way to format an event as text suitable for a human being to read, via the function :api:`twisted.logger.formatEvent <formatEvent>` .
+Format strings provide observers with a standard way to format an event as text suitable for a human being to read.  Formatting is accomplished using the functions :api:`twisted.logger.formatEvent <formatEvent>` and :api:`twisted.logger.formatEventWithTracebck <formatEventWithTraceback>`.  The latter should be preferred unless potential traceback information in the event is not important.
```

Comment:
I don't love the fact that we now have 3 subtly different formatting functions (`formatEvent`, `formatEventWithTraceback`, `formatEventAsClassicLogText`).  These are separate APIs just to preserve the compatibility contract; it's not a thoughtful API surface.

However, in addition to new public API, this is also a serious bugfix, so I don't want to waffle on it much longer.

What I think I'd like to see in the future is a single API: `eventAsText`, which takes

- event: the event, obviously
- includeTraceback=True: include a traceback if `log_failure`
- includeTimestamp=True: include a prefixed timestamp
- includeSystem=True: include the `[system]` string before the message, but after the timestamp if included

`formatEvent`, `formatEventWithTraceback`, and `formatEventAsClassicLogText` would then just become aliases for various combinations of parameters to this function, and we could just document this function.

We *could* just skip straight to that endpoint if you wanted to do extra work right now though.

Reply:
I agree with your comments on the thoughtfulness of the API.  For personal context, backward compatibility is something that is extremely important to me (code I wrote against 13.x hasn't needed much updating to run on 17.9, that is *amazing*, kudos to the years of hard work in Twisted), so when I make changes such as this I always err on the side of compatibility...unfortunately, that can lead to this kind of mess.

@glyph I'm in no rush, I'd rather see a better public API if maintainers are open to it, and I like what you've suggested above.  One change to your proposal I suggest is, since `formatEventWithTraceback` was a new API proposed in this PR, it needn't exist as an alias if eventAsText exists with the semantics above.  The other APIs are legacy and as such should be preserved.

I'd also like to fix 9456 as well, if there are no objections?

Do you want a new ticket for all of this or can I just use 7927?

Reply:
Sounds like we're very much in sync on the compatibility / thoughtfulness stuff.  I'd like to get this landed soon, since we've got lots of great traceback changes coming in 18.7, and I'd really like to resolve this additional annoyance quickly following on the heels of that release, but I don't want to rush it especially if you're available to respond to the feedback promptly.

Smaller PRs are better, but, I don't think 9456 would substantially increase the size of this one, and it would be nice to close more tickets, so feel free to pull it in here.

Reply:
I'm also going to throw a mention for @wsanchez on here since this API involved a lot of feedback from him and if we're lucky he'll have a minute to look at the suggestion of `eventAsText` and see if there are any corner cases I've forgotten.

Reply:
@glyph I should be able to get to it fairly quickly.  Obviously things come up, so there's an element out of my hands, but barring that I'd expect the changes soon.  A procedure question, where this is moving in a different direction -- do you want me to use the same PR or start fresh with a new branch and clean history?

Reply:
Dealer's choice - just make it clear on the ticket where the code is when you put it back into review.